algorithms/scripts/RIT1-PCA+LDA-htcai.py

```diff
@@ -0,0 +1,219 @@
+    'l1_ratio': [0, 0.2, 0.5, 0.8, 1],
+}
+
+
+# In[14]:
+
+get_ipython().run_cell_magic('time', '', "clf = SGDClassifier(random_state=0, class_weight='balanced', loss='log', penalty='elasticnet')\ncv = GridSearchCV(estimator=clf, param_grid=param_grid, n_jobs=-1, scoring='roc_auc')\ncv.fit(X = X_train_lda, y=y_train)")
+
+
+# In[15]:
+
+# Best Params
+print('{:.3%}'.format(cv.best_score_))
+
+# Best Params
+cv.best_params_
+
+
+# ## Visualize hyperparameters performance
+
+# In[16]:
+
+cv_result_df = pd.concat([
+    pd.DataFrame(cv.cv_results_),
+    pd.DataFrame.from_records(cv.cv_results_['params']),
+], axis='columns')
+cv_result_df.head(2)
+
+
+# In[17]:
+
+# Cross-validated performance heatmap
+cv_score_mat = pd.pivot_table(cv_result_df, values='mean_test_score', index='l1_ratio', columns='alpha')
+ax = sns.heatmap(cv_score_mat, annot=True, fmt='.1%')
+ax.set_xlabel('Regularization strength multiplier (alpha)')
+ax.set_ylabel('Elastic net mixing parameter (l1_ratio)');
+
+
+# ## Use Optimal Hyperparameters to Output ROC Curve
+
+# In[18]:
+
+y_pred_train = cv.decision_function(X_train_lda)
+y_pred_test = cv.decision_function(X_test_lda)
+
+def get_threshold_metrics(y_true, y_pred):
+    roc_columns = ['fpr', 'tpr', 'threshold']
+    roc_items = zip(roc_columns, roc_curve(y_true, y_pred))
+    roc_df = pd.DataFrame.from_items(roc_items)
+    auroc = roc_auc_score(y_true, y_pred)
+    return {'auroc': auroc, 'roc_df': roc_df}
+
+metrics_train = get_threshold_metrics(y_train, y_pred_train)
+metrics_test = get_threshold_metrics(y_test, y_pred_test)
+
+
+# In[19]:
+
+# Plot ROC
+plt.figure()
+for label, metrics in ('Training', metrics_train), ('Testing', metrics_test):
+    roc_df = metrics['roc_df']
+    plt.plot(roc_df.fpr, roc_df.tpr,
+        label='{} (AUROC = {:.1%})'.format(label, metrics['auroc']))
+plt.xlim([0.0, 1.0])
+plt.ylim([0.0, 1.05])
+plt.xlabel('False Positive Rate')
+plt.ylabel('True Positive Rate')
+plt.title('Predicting RIT1 mutation from gene expression (ROC curves)')
+plt.legend(loc='lower right');
+
+
+# ## Investigate the predictions
+
+# In[20]:
+
+X_transformed = lda.transform(pca.transform(X))
+
+
+# In[21]:
+
+predict_df = pd.DataFrame.from_items([
+    ('sample_id', X.index),
+    ('testing', X.index.isin(X_test.index).astype(int)),
+    ('status', y),
+    ('decision_function', cv.decision_function(X_transformed)),
+    ('probability', cv.predict_proba(X_transformed)[:, 1]),
+])
+predict_df['probability_str'] = predict_df['probability'].apply('{:.1%}'.format)
+
+
+# In[22]:
+
+# Top predictions amongst negatives (potential hidden responders)
+predict_df.sort_values('decision_function', ascending=False).query("status == 0").head(10)
+
+
+# In[23]:
+
+# Ignore numpy warning caused by seaborn
+warnings.filterwarnings('ignore', 'using a non-integer number instead of an integer')
+
+ax = sns.distplot(predict_df.query("status == 0").decision_function, hist=False, label='Negatives')
```

Comment:
It looks like the LDA transform causes the decision function to shift far to the negative. Any reason for this?